transport/http/middleware_content_length.go

```diff
@@ -0,0 +1,50 @@
+package http
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/awslabs/smithy-go/middleware"
+)
+
+// ContentLengthMiddleware provides a middleware to set the content-length
+// header for the length of a serialize request body.
+type ContentLengthMiddleware struct {
+}
+
+// AddContentLengthMiddleware adds ContentLengthMiddleware to the middleware
+// stack's Build step.
+func AddContentLengthMiddleware(stack *middleware.Stack) {
+	stack.Build.Add(&ContentLengthMiddleware{}, middleware.After)
+}
+
+// ID the identifier for the ContentLengthMiddleware
+func (m *ContentLengthMiddleware) ID() string { return "ContentLengthMiddleware" }
+
+// HandleBuild adds the length of the serialized request to the HTTP header
+// if the length can be determined.
+func (m *ContentLengthMiddleware) HandleBuild(
+	ctx context.Context, in middleware.BuildInput, next middleware.BuildHandler,
+) (
+	out middleware.BuildOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := in.Request.(*Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown request type %T", req)
+	}
+
+	// Don't set content length if header is already set.
+	if vs := req.Header.Values("Content-Length"); len(vs) != 0 {
```

Comment:
We have an open ticket that says Go SDK will include a ```content-type``` even when ```Content-Length=0``` in header. 

If it's the case, should we remove that header here?

Reply:
Will take a look at this. Removing `Content-Type` probably should be its own middleware. Its kind of contextual though, because an for example its value for a S3 PutObject operation call to send a Content-Type=text/plain for a zero length object.  Maybe we should only remove 


Though this does remind me that the SDK's signer must not include `Content-Length` as a signed header if the value is `0` because intermittent network hops can remove the header if its value is zero.

Reply:
Created https://github.com/awslabs/smithy-go/issues/110 to track removing Content-Type when length is 0.